### PR TITLE
[TDF] Fix ROOT-9218

### PR DESCRIPTION
**Pull request by dpiparo**
Checking the type of the columns also when provided by a data source.

### CODE REVIEW

File: tree/treeplayer/src/TDFUtils.cxx [hunk 1]
```diff
@@ -36,6 +36,84 @@ TIgnoreErrorLevelRAII::~TIgnoreErrorLevelRAII()
...
+      msg += ".";
+      throw std::runtime_error(msg);
+   }
+}
+
+std::string TypeID2TypeName(const std::type_info &id)
+{
+   if (auto c = TClass::GetClass(id)) {
+      return c->GetName();
+   } else if (type_id == typeid(char))
```
Comment (pcanal) [inline tree/treeplayer/src/TDFUtils.cxx]: See also TDataType::GetDataType

File: tree/treeplayer/src/TDFUtils.cxx [hunk 2]
```diff
@@ -36,6 +36,79 @@ TIgnoreErrorLevelRAII::~TIgnoreErrorLevelRAII()
...
+      return typeid(long);
+   else if (name == "unsigned long" || name == "ULong_t")
+      return typeid(unsigned long);
+   else if (name == "double" || name == "Double_t")
+      return typeid(double);
+   else if (name == "float" || name == "Float_t")
+      return typeid(float);
+   else if (name == "Long64_t")
+      return typeid(Long64_t);
+   else if (name == "ULong64_t")
```
Comment (bluehood) [inline tree/treeplayer/src/TDFUtils.cxx]: `long long` and `unsigned long long` should probably be added next to Long64_t and ULong64_t

File: tree/treeplayer/src/TDFUtils.cxx [hunk 3]
```diff
@@ -36,6 +36,79 @@ TIgnoreErrorLevelRAII::~TIgnoreErrorLevelRAII()
    gErrorIgnoreLevel = fCurIgnoreErrorLevel;
 }
 
+const std::type_info &TypeName2TypeID(const std::string &name)
```
Comment (bluehood) [inline tree/treeplayer/src/TDFUtils.cxx]: maybe we should leave a comment for posterity explaining that we do not deal with pointer/reference types because TDF does not accept columns of pointer type and references are removed before getting here

File: tree/treeplayer/src/TDFUtils.cxx [hunk 4]
```diff
@@ -36,6 +36,79 @@ TIgnoreErrorLevelRAII::~TIgnoreErrorLevelRAII()
...
+      return typeid(bool);
+   else {
+      std::string msg("Cannot extract type_info of type ");
+      msg += name.c_str();
+      msg += ".";
+      throw std::runtime_error(msg);
+   }
+}
+
+std::string TypeID2TypeName(const std::type_info &id)
```
Comment (bluehood) [inline tree/treeplayer/src/TDFUtils.cxx]: as above, maybe we should leave a comment for posterity explaining that we do not deal with pointer/reference types because TDF does not accept columns of pointer type and references are removed before getting here

File: tree/treeplayer/src/TDFUtils.cxx [hunk 5]
```diff
@@ -77,48 +150,23 @@ ColumnName2ColumnTypeName(const std::string &colName, TTree *tree, TCustomColumn
             return l->GetTypeName();
          } else {
             // we do not know how to deal with this branch
-            throw std::runtime_error("TTree branch " + colName + " has both a leaf count and a static length. This is not supported.");
+            throw std::runtime_error("TTree branch " + colName +
+                                     " has both a leaf count and a static length. This is not supported.");
```
Comment (bluehood) [inline tree/treeplayer/src/TDFUtils.cxx]: formatting unrelated lines might make it harder to cherry-pick this commit -- but it's my fault for not formatting this line properly before 😅 I wonder why travis let this pass

File: tree/treeplayer/src/TRootDS.cxx [hunk 6]
```diff
@@ -13,8 +13,19 @@ namespace ROOT {
...
+std::vector<void *> TRootDS::GetColumnReadersImpl(std::string_view name, const std::type_info &id)
 {
+   const auto colTypeName = GetTypeName(name);
+   const auto &colTypeId = ROOT::Internal::TDF::TypeName2TypeID(colTypeName);
+   if (id != colTypeId) {
+      std::string err = "The type of column \"";
+      err += name;
+      err += "\" is ";
+      err += colTypeName;
+      err += " but a different one has been selected.";
```
Comment (bluehood) [inline tree/treeplayer/src/TRootDS.cxx]: maybe `id.name()` would give users an extra hint of what they did wrong -- or maybe it would confuse them more 😅
